platforms/nuttx/src/px4_layer/console_buffer.cpp

```diff
@@ -0,0 +1,160 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2019 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include <px4_config.h>
+#include <px4_console_buffer.h>
+#include <px4_defines.h>
+#include <px4_sem.h>
+
+#include <fcntl.h>
+
+#ifdef CONFIG_ARCH_BOARD_PX4_FMU_V5
+
+static ssize_t console_buffer_write(struct file *filep, const char *buffer, size_t buflen);
+
+#define CONSOLE_BUFFER_LENGTH (1024*4)
+
+class ConsoleBuffer
+{
+public:
+
+	void write(const char *buffer, size_t len);
+
+	void print(bool follow);
+
+private:
+	void		lock() { do {} while (px4_sem_wait(&_lock) != 0); }
+	void		unlock() { px4_sem_post(&_lock); }
+
+	char _buffer[CONSOLE_BUFFER_LENGTH];
+	int _head{0};
+	int _tail{0};
+	px4_sem_t _lock = SEM_INITIALIZER(1);
+	bool _is_printing{false};
+};
+
+void ConsoleBuffer::print(bool follow)
+{
+	lock();
+	int i = _head;
+
+	while (true) {
+		_is_printing = true;
+
+		for (; i != _tail; i = (i + 1) % CONSOLE_BUFFER_LENGTH) {
+			putchar(_buffer[i]);
+		}
+
+		_is_printing = false;
+
+		if (follow) {
+			unlock();
+			usleep(10000);
+			lock();
+
+		} else {
+			break;
+		}
+	}
+
+	unlock();
+}
+
+void ConsoleBuffer::write(const char *buffer, size_t len)
+{
+	if (_is_printing) { // avoid adding to the buffer while we are printing it
```

Comment:
Hm, so we potentially miss things. And you don't want to lock and wait until writing is done?

Reply:
True, there's a minimal chance for that if one thread is printing and another does a write.
However we cannot take the lock here, otherwise we deadlock.

Reply:
Ok, why do we deadlock?

Reply:
Fixed.